File: rust/libnewsboat-ffi/src/filepath.rs
use cxx::{type_id, ExternType};
use std::ffi::OsStr;
use std::os::unix::ffi::OsStrExt;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
pub struct PathBuf(pub std::path::PathBuf);

unsafe impl ExternType for PathBuf {
    type Id = type_id!("newsboat::filepath::bridged::PathBuf");
    type Kind = cxx::kind::Opaque;
}

#[cxx::bridge(namespace = "newsboat::filepath::bridged")]
mod bridged {
    extern "Rust" {
        type PathBuf;

        fn create_empty() -> Box<PathBuf>;
        fn create(filepath: Vec<u8>) -> Box<PathBuf>;
        fn equals(lhs: &PathBuf, rhs: &PathBuf) -> bool;
        fn less_than(lhs: &PathBuf, rhs: &PathBuf) -> bool;
        fn into_bytes(filepath: &PathBuf) -> Vec<u8>;
        fn display(filepath: &PathBuf) -> String;
        fn push(filepath: &mut PathBuf, component: &PathBuf);
        fn clone(filepath: &PathBuf) -> Box<PathBuf>;
        fn is_absolute(filepath: &PathBuf) -> bool;
        fn set_extension(filepath: &mut PathBuf, extension: Vec<u8>) -> bool;
        fn starts_with(filepath: &PathBuf, base: &PathBuf) -> bool;
        fn file_name(filepath: &PathBuf) -> Vec<u8>;
    }
}

fn create_empty() -> Box<PathBuf> {
    Box::new(PathBuf(std::path::PathBuf::new()))
}

fn create(filepath: Vec<u8>) -> Box<PathBuf> {
    let filepath: &OsStr = OsStrExt::from_bytes(&filepath);
    let filepath = std::path::Path::new(filepath);
    Box::new(PathBuf(filepath.to_path_buf()))
}

fn equals(lhs: &PathBuf, rhs: &PathBuf) -> bool {
    lhs.0 == rhs.0
}

fn less_than(lhs: &PathBuf, rhs: &PathBuf) -> bool {
    lhs.0 < rhs.0
}

fn into_bytes(filepath: &PathBuf) -> Vec<u8> {
    filepath.0.as_os_str().as_bytes().to_owned()
}

fn display(filepath: &PathBuf) -> String {
    format!("{}", filepath.0.display())
}

fn push(filepath: &mut PathBuf, component: &PathBuf) {
    filepath.0.push(&component.0)
}

fn clone(filepath: &PathBuf) -> Box<PathBuf> {
    Box::new(PathBuf(filepath.0.clone()))
}

fn is_absolute(filepath: &PathBuf) -> bool {
    filepath.0.is_absolute()
}

fn set_extension(filepath: &mut PathBuf, extension: Vec<u8>) -> bool {
    filepath.0.set_extension(OsStr::from_bytes(&extension))
}

fn starts_with(filepath: &PathBuf, base: &PathBuf) -> bool {
    filepath.0.starts_with(&base.0)
}

fn file_name(filepath: &PathBuf) -> Vec<u8> {
    if let Some(res) = filepath.0.file_name() {
        res.as_bytes().to_vec()
    } else {
        Vec::new()
    }
}

File: src/filepath.cpp
#define ENABLE_IMPLICIT_FILEPATH_CONVERSIONS

#include "filepath.h"

namespace newsboat {

inline namespace {

rust::Vec<std::uint8_t> string_to_vec(const std::string& input)
{
	rust::Vec<std::uint8_t> result;
	result.reserve(input.length());
	for (const auto byte : input) {
		result.push_back(byte);
	}
	return result;
}

}

Filepath::Filepath()
	: rs_object(filepath::bridged::create_empty())
{
}

Filepath Filepath::from_locale_string(const std::string& filepath)
{
	Filepath result;
	result.rs_object = filepath::bridged::create(string_to_vec(filepath));
	return result;
}

Filepath::Filepath(const Filepath& filepath)
	: rs_object(filepath::bridged::clone(*filepath.rs_object))
{
}

Filepath& Filepath::operator=(const Filepath& filepath)
{
	if (this == &filepath) {
		return *this;
	}
	rs_object = filepath::bridged::clone(*filepath.rs_object);
	return *this;
}

std::string Filepath::to_locale_string() const
{
	const auto bytes = filepath::bridged::into_bytes(*rs_object);
	return std::string(std::begin(bytes), std::end(bytes));
}

std::string Filepath::display() const
{
	return std::string(filepath::bridged::display(*rs_object));
}

bool Filepath::operator==(const Filepath& other) const
{
	return filepath::bridged::equals(*rs_object, *other.rs_object);
}

bool Filepath::operator!=(const Filepath& other) const
{
	return !(*this == other);
}

void Filepath::push(const Filepath& component)
{
	filepath::bridged::push(*rs_object, *component.rs_object);
}

Filepath Filepath::join(const Filepath& component) const
{
	auto result = *this;
	result.push(component);
	return result;
}

bool Filepath::is_absolute() const
{
	return filepath::bridged::is_absolute(*rs_object);
}

bool Filepath::set_extension(const std::string& ext)
{
	return filepath::bridged::set_extension(*rs_object, string_to_vec(ext));
}

bool Filepath::starts_with(const std::string& base) const
{
	return filepath::bridged::starts_with(*rs_object, string_to_vec(base));
}

std::optional<Filepath> Filepath::file_name() const
{
	auto str = filepath::bridged::file_name(*rs_object);
	auto res = std::string(str.begin(), str.end());
	if (res.empty()) {
		return std::nullopt;
	} else {
		return res;
	}
}

} // namespace newsboat
